Userland/Tests/Kernel/kill-pidtid-confusion.cpp

```diff
@@ -0,0 +1,171 @@
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include <AK/Assertions.h>
+#include <AK/LogStream.h>
+#include <LibPthread/pthread.h>
+#include <fcntl.h>
+#include <signal.h>
+#include <stdio.h>
+#include <unistd.h>
+
+/*
+ * Bug:
+ * If the main thread of a process is no longer alive, it cannot receive
+ * signals anymore. This can manifest as, for example, an unkillable process.
+ *
+ * So what needs to happen:
+ * - There is process P
+ * - It has more than one thread
+ * - The main thread calls thread_exit(), leaving the rest of the threads alive
+ * - Now the process is unkillable!
+ *
+ * Here's how to demonstrate the bug:
+ * - Time 0: PX forks into PZ (mnemonic: Zombie)
+ * - Time 1: PZ's main thread T1 creates a new thread T2
+ * - Time 2: Nothing (T2 could communicate to PX both process and thread ID)
+ *      (most LibC functions crash currently, which is a different bug I suppose.)
+ * - Time 3: T1 calls thread_exit()
+ * - Time 4:
+ *      * PX tries to kill PZ (should work, but doesn't)
+ *      * PX tries to kill PZ using T2's thread ID (shouldn't work, and doesn't)
+ *      * PX outputs all results.
+ */
+
+static constexpr useconds_t STEP_SIZE = 1100000;
+
+static void fork_into(void(fn)())
+{
+    const pid_t rc = fork();
+    if (rc < 0) {
+        perror("fork");
+        exit(1);
+    }
+    if (rc > 0) {
+        return;
+    }
+    fn();
+    dbg() << "child finished (?)";
+    exit(1);
+}
+
+static void thread_into(void* (*fn)(void*))
+{
+    pthread_t tid;
+    const int rc = pthread_create(&tid, nullptr, fn, nullptr);
+    if (rc < 0) {
+        perror("pthread_create");
+        exit(1);
+    }
+}
+
+static void sleep_steps(useconds_t steps)
+{
+    const int rc = usleep(steps * STEP_SIZE);
+    if (rc < 0) {
+        perror("usleep");
+        ASSERT_NOT_REACHED();
+    }
+}
+
+static bool try_kill(pid_t kill_id)
+{
+    int rc = kill(kill_id, SIGTERM);
+    perror("kill");
+    printf("kill rc: %d\n", rc);
+    return rc >= 0;
+}
+
+static void run_pz();
+static void* run_pz_t2_wrap(void* fd_ptr);
+static void run_pz_t2();
+
+int main(int, char**)
+{
+    // This entire function is the entirety of process PX.
+
+    // Time 0: PX forks into PZ (mnemonic: Zombie)
+    // Serenity doesn't support O_NONBLOCK for pipes yet, so
+    // sadly the test will hang if something goes wrong.
```

Comment:
Hmm, `O_NONBLOCK` is implemented at the syscall level (read/write/etc), not at the File subclass level. What's the issue here about?

Reply:
That comment is outdated, from a time when I tried to communicate pid/tid of "T2" back to the main process through a pipe. I abandoned that approach, as it was too much work, and also currently there's some other bug that makes any call to `dbg()`, `printf()`, `write()`, and probably most other things crash the process (here: PZ) under those circumstances. And I wanted to make the test work *now*, so that it can show *now* what the difference is.

So, in short: outdated comment, will remove.